fw/access_log.c

```diff
@@ -0,0 +1,430 @@
+	 * which should be always false */
+	BUG_ON(end - p < TFW_ADDR_STR_BUF_SIZE);
+#define FMT_client_ip "%.*s"
+#define ARG_client_ip , client_ip.len, client_ip.data
+	if (req->conn && req->conn->peer) {
+		client_ip.data = p;
+		p = tfw_addr_fmt(&req->conn->peer->addr, TFW_NO_PORT, p);
+		client_ip.len = p - client_ip.data;
+	} else {
+		client_ip = missing;
+	}
+	/* vhost */
+#define FMT_vhost "%.*s"
+#define ARG_vhost , vhost.len, vhost.data
+	if (req->vhost && !TFW_STR_EMPTY(&req->vhost->name)) {
+		p = make_plain(p, end, &req->vhost->name, &vhost);
+	} else {
+		vhost = missing;
+	}
+	/* method */
+#define FMT_method "%.*s"
+#define ARG_method , method.len, method.data
+	if (req->method < sizeof(http_methods) / sizeof(*http_methods)
+			&& http_methods[req->method].len != 0) {
+		method.data = http_methods[req->method].name;
+		method.len = http_methods[req->method].len;
+	} else {
+		method = missing;
+	}
+	/* http version */
+#define FMT_version "%.*s"
+#define ARG_version , version.len, version.data
+	if (req->version < sizeof(http_versions) / sizeof(*http_versions)
+			&& http_versions[req->version].len != 0) {
+		version.data = http_versions[req->version].name;
+		version.len = http_versions[req->version].len;
+	} else {
+		version = missing;
+	}
+	/* status, content_length */
+	/* NOTE: we only roughly estimate lengths of numbers, leaving final
+	 * transformation to printk. This has some side-effects like string
+	 * will be truncated while being smaller that destination buffer, but
+	 * that would do for a while */
+#define FMT_status "%d"
+#define ARG_status , resp_status
+#define FMT_content_length "%lu"
+#define ARG_content_length , resp_content_length
+	status.data = "";
+	status.len = 10; /* len(str(2**32)) */
+	content_length.data = "";
+	content_length.len = 20; /* len(str(2**64)) */
+	/* Process truncated fields */
+	truncated_in[idx_uri] = req->uri_path;
+#define ADD_HDR(id, tfw_hdr_id)                                      \
+		truncated_in[id] = get_http_header_value(req->version, \
+				req->h_tbl->tbl + tfw_hdr_id);
+	ADD_HDR(idx_referer, TFW_HTTP_HDR_REFERER);
+	ADD_HDR(idx_user_agent, TFW_HTTP_HDR_USER_AGENT);
+	/* Now we calculate first estimation of
+	 * "maximum allowed truncated string length" */
+#define ESTIMATE_FIXED(str) + (sizeof(str) - 1)
+#define ESTIMATE_UNTRUNCATABLE(id) + id.len
+#define ESTIMATE_TRUNCATABLE(id)
+	process_truncated(truncated_in, truncated_out, p, end,
+			ACCESS_LOG_LINE(ESTIMATE_FIXED, ESTIMATE_UNTRUNCATABLE,
+					ESTIMATE_TRUNCATABLE));
+	/* Use macro to build format string */
+#define FMT_FIXED(str) str
+#define FMT_UNTRUNCATABLE(id) FMT_ ## id
+#define FMT_TRUNCATABLE(id) "%.*s"
+#define ARG_FIXED(str)
+#define ARG_UNTRUNCATABLE(id) ARG_ ## id
+#define ARG_TRUNCATABLE(id) , truncated_out[idx_ ## id].len, \
+		truncated_out[idx_ ## id].data
+#ifdef XXX
+BEGIN
+#endif
+
+	/* we have to enforce some macro expansion here */
+#define DO_PR_INFO(...) pr_info(__VA_ARGS__)
```

Comment:
We don't need to define a macro just for one usage. The code is already overloaded with macros and it'd be good to reduce the number of macros

Reply:
I've added explaining comment. The macro could be more generic, but it is absolutely necessary to expand arguments to invoke another macro. This is required only because `pr_info` is a macro itself.